rq/worker.py

```diff
@@ -1573,3 +1584,208 @@ class RandomWorker(Worker):
+        When all queues are empty, block and wait for new jobs to arrive on any of the
+        queues, unless `burst` mode is enabled.
+
+        The return value indicates whether any jobs were processed.
+        """
+        self.bootstrap(logging_level, date_format, log_format)
+        completed_jobs = 0
+        self.log.info("ThreadPoolWorker %s: started with %s threads, version %s", self.key, self.threadpool_size, VERSION)
+        self.log.warning("*** WARNING: ThreadPoolWorker is in beta and may be unstable. Don't use it in production!")
+        if with_scheduler:
+            self._start_scheduler(burst, logging_level, date_format, log_format)
+
+        self._install_signal_handlers()
+        try:
+            while True:
+                try:
+                    self.check_for_suspension(burst)
+
+                    if self.should_run_maintenance_tasks:
+                        self.run_maintenance_tasks()
+
+                    if self._stop_requested:
+                        self.log.info('Worker %s: stopping on request', self.key)
+                        break
+
+                    if self.is_pool_full:
+                        self.log.debug('ThreadPool is full, waiting for idle threads...')
+                        self.__wait_for_slot()
+
+                    timeout = None if burst else self.dequeue_timeout
+                    result = self.dequeue_job_and_maintain_ttl(timeout)
+                    if result is None:
+                        if not burst:
+                            break
+                        has_pending_dependents = self.__check_pending_dependents()
+                        if has_pending_dependents:
+                            continue
+                        self.log.info("Worker %s: done, quitting", self.key)
+                        break
+
+                    job, queue = result
+                    self.reorder_queues(reference_queue=queue)
+                    self.execute_job(job, queue)
+                    self.heartbeat()
+
+                    completed_jobs += 1
+                    if max_jobs is not None:
+                        if completed_jobs < max_jobs:
+                            continue
+                        self.log.info("Worker %s: finished executing %d jobs, quitting", self.key, completed_jobs)
+                        break
+
+                except redis.exceptions.TimeoutError:
+                    self.log.error(f"Worker {self.key}: Redis connection timeout, quitting...")
+                    break
+
+                except StopRequested:
+                    break
+
+                except SystemExit:
+                    # Cold shutdown detected
+                    raise
+
+                except:  # noqa
+                    self.log.error('Worker %s: found an unhandled exception, quitting...', self.key, exc_info=True)
+                    break
+        finally:
+            self.register_death()
+            self.unsubscribe()
```

Comment:
These two lines can go into `worker.teardown()`

Reply:
We don't have this method now  - you mean creating a new method on the main worker?

Reply:
Added on #1820